### ISSUE TITLE
CombinedResourceHandler: Add subresource integrity attribute

### ISSUE THREAD

**Issue by melloware** [omnifaces/omnifaces#660]
SubResource Integrity Spec: https://www.w3.org/TR/SRI/

To add the SHA384 integrity and `crossorigin="anonymous"` attributes to the combined CSS and JS links like...

```xml
<link type="text/css" 
   rel="stylesheet" 
   href="/javax.faces.resource/eNorKMrMTU1LTE4ttkrOzy3Iz0vNKynWSy4urslJrMwvLbECMvUhTLBocUZ-eXJicSpYHMbBlCkoSi0pyUyrBMkAAM-bJ9A.css.xhtml?ln=omnifaces.combined&amp;v=1639562619188"
   integrity="sha384-+/M6kredJcxdsqkczBUjMLvqyHb1K/JThDXWsBVxMEeZHEaMKEOEct339VItX1zB"
   crossorigin="anonymous" />
```

**Comment by melloware** [comment]
Something like this should do it...

```java
public static String calculateIntegrity(String input)
    {
        try {
            // getInstance() method is called with algorithm SHA-384
            MessageDigest md = MessageDigest.getInstance("SHA-384");
  
            // digest() method is called
            // to calculate message digest of the input string
            // returned as array of byte
            byte[] messageDigest = md.digest(input.getBytes());
  
            // Convert byte array into signum representation
            BigInteger no = new BigInteger(1, messageDigest);
  
            // Convert message digest into hex value
            String hashtext = no.toString(16);
  
            // Add preceding 0s to make it 32 bit
            while (hashtext.length() < 32) {
                hashtext = "0" + hashtext;
            }

           // now base64 encode the hash
           String encodedString = Base64.getEncoder().encodeToString(hashtext.getBytes());
  
            // return the HashText
            return encodedString ;
        }
  
        // For specifying wrong message digest algorithms
        catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

```

**Comment by BalusC** [comment]
Implemented for 3.13.

Can you please confirm if https://oss.sonatype.org/content/repositories/snapshots/org/omnifaces/omnifaces/3.13-SNAPSHOT/ works for you?

**Comment by melloware** [comment]
Awesome confirmed its working and that the browser is reporting the SHA is correct for the JS and CSS being uploaded!!!

Now I have to fix a bug in PrimeFaces MOVE_SCRIPTS_TO_BOTTOM to make this work for me too: https://github.com/primefaces/primefaces/issues/7882

😄

**Comment by melloware** [comment]
PF is fixed for 12.0!  i think all around this improves the security of JSF app.  Thanks for the quick turnaround.

**Comment by BalusC** [comment]
We need to rethink this. The integrity attribute is only mandatory for cross origin resources not for local resources. See also https://github.com/omnifaces/omnifaces/issues/876
